### PR TITLE
Rewrite and open source wgtunnel

### PR DESCRIPTION
**Pull request by deansheather**
Rewrite and open source our `wgtunnel` server used by [coder/coder](https://github.com/coder/coder).

Compared to our closed source version:
- the code is more tidy and separated
- it contains a client implementation
- it contains tests

## TODO:

- [x] Add back tracing
- [x] Fix the tunnel so tests pass
- [x] Add CI for tests
- [x] Add CI for releases
- [x] Add signal handling
- [x] Print the tunnel URL in `cmd/tunnel`
- [ ] Add legacy hostname support like old wgtunnel (in future PR)

### CODE REVIEW

File: tunneld/tunneld_test.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+		Log: slogtest.
+			Make(t, &slogtest.Options{IgnoreErrors: true}).
+			Named("tunnel_client"),
+		PrivateKey: key,
+	})
+	require.NoError(t, err, "launch tunnel")
+	t.Cleanup(func() {
+		_ = tunnel.Close()
+		<-tunnel.Wait()
+	})
```
Comment (coadler) [inline tunneld/tunneld_test.go:60]: Personally I still think defers make sense for this. `t.Cleanup` should be used for when stuff should be cleaned up outside of the current function.

File: tunnelsdk/client.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+// Client provides HTTP methods for the tunneld API and a full wireguard tunnel
+// client implementation.
+type Client struct {
+	HTTPClient *http.Client
+	URL        *url.URL
+}
+
+// Request performs an HTTP request with the body provided. The caller is
+// responsible for closing the response body.
+func (c *Client) Request(ctx context.Context, method, path string, body interface{}) (*http.Response, error) {
```
Comment (coadler) [inline tunnelsdk/client.go:35]: I think this should just be unexported.

File: tunnelsdk/tunnel.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+type TunnelConfig struct {
+	Log slog.Logger
+	// PrivateKey is the Wireguard private key. You can use GeneratePrivateKey
+	// to generate a new key. It should be stored in a safe place for future
+	// tunnel sessions, otherwise you will get a new hostname.
+	PrivateKey Key
+}
+
+func (c *Client) LaunchTunnel(ctx context.Context, cfg TunnelConfig) (*Tunnel, error) {
```
Comment (coadler) [inline tunnelsdk/tunnel.go:122]: This should probably have a comment.

**Review by coadler — APPROVED** [review]

File: LICENSE [hunk 4]
```diff
@@ -0,0 +1,661 @@
+                    GNU AFFERO GENERAL PUBLIC LICENSE
```
Comment (deansheather) [inline LICENSE:1]: @ghuntley thinks we should change this to MIT

File: README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+```console
+$ go install github.com/coder/wgtunnel/cmd/tunnel
+```
+
+or by running `make build/tunnel`.
+
+## License
+
+Licensed under the AGPL-3.0 license.
```
Comment (ghuntley) [inline README.md:49]:
```suggestion
Licensed under the MIT license.
```

File: .github/workflows/release.yaml [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+          echo "docker_image_tag=$(cat build/tunneld.tag)" >> $GITHUB_ENV
+
+      - name: Push Docker image
+        if: ${{ !github.event.inputs.dry_run && !github.event.inputs.snapshot }}
+        run: |
+          set -euxo pipefail
+
+          docker push "${{ env.docker_image_tag }}"
+
+          latest_tag="ghcr.io/wgtunnel/wgtunnel/tunneld:latest"
```
Comment (deansheather) [inline .github/workflows/release.yaml:70]: incorrect URL

**Pull request merged**